dts/msm8974/msm8974-samsung-kltechn.dts

```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/dts-v1/;
+
+#include <skeleton.dtsi>
+
+/ {
+	// This is used by the bootloader to find the correct DTB
+	qcom,msm-id = <0xC208FF01 10 0x10000>;
+
+	kltechn {
+		model = "Samsung Galaxy S5 China (SM-G9006V/8V) / China Duos (SM-G9006W/8W/9W)";
+		compatible = "samsung,klte", "qcom,msm8974", "lk2nd,device";
+		lk2nd,match-bootloader = "G900*";
```

Comment:
> maybe better to have the G9006* and leave other variants out for now?

I am afraid this is not suitable. In China, the most popular model is `SM-G9008V/W`, in the meantime, other models are not rare. They are all carrier-limited variants: `G9006` for China Unicom, `G9008` for China Mobile, and `G9009` for China Telecom. Besides the carrier limit, they are totally the same.

Matching `G900*` indeed could potentially lead to some issues, but I just did not find a better solution. Maybe an improvement on the string matcher is needed?

In fact, even changing it to `G900F*` (which is absolutely wrong) won't break anything...

Reply:
I assume there is also `G9000` which is not carrier locked? I think there would need to be a node for each carrier variant then but other ones bootloaders might want to select a different revision dtb (who knows how sammy messed with them, can see how many revisions we track in the msm8916 dir) so adding nodes in advance is not something I'd prefer to do...

Reply:
> I assume there is also G9000 which is not carrier locked?

Nope. I've never heard of that and after doing a [search](https://www.samsung.com/hk_en/search/?searchvalue=g9000) I am sure that there is not such a model. Galaxy S5 was released in 2014 and that was not a carrier-lock-free age.
I've done some research and found that all models matching `G900(\d\w*)` are [`6V/W`](https://www.samsung.com/hk_en/search/?searchvalue=g9006), [`8V/W`](https://www.samsung.com/hk_en/search/?searchvalue=g9008), and [`9W/D`](https://www.samsung.com/hk_en/search/?searchvalue=g9009). They are all Chinese variants with carrier limits. `\dV` means single SIM slot (`kltechn`), `\dW` means duo SIM slots (`kltechnduo`), and `\dD` means duo SIM slots w/o LTE support (`k3gduosctc`).

> I think there would need to be a node for each carrier variant then

Since they share the same dts, I hardly get your point.
Distinguishing `G9009D` (`qcom,msm-id = <0xD808FF01 14 0x10000>;`, [dts](https://github.com/LineageOS/android_kernel_samsung_msm8974/blob/lineage-18.1/arch/arm/boot/dts/msm8974pro/msm8974pro-ac-sec-kctc-r14.dts)) from the others should be enough, and it already has a different `chipset_foundry_id`.
The official Android OS seems to be the **only** OS that needs to distinguish those carrier variants. Third-party OSes just need to distinguish `kltechnduo` from `kltechn`. So could you please explain your point more detailedly?

> other ones bootloaders might want to select a different revision dtb

If I did not misunderstand, the "revision" should be **hardware** revision thus selecting a different revision dtd should be intolerable. Any retailed device should be at the latest revision and the old revisions are just some development legacy not deleted.

> adding nodes in advance is not something I'd prefer to do...

I see.

Reply:
> I assume there is also `G9000` which is not carrier locked? I think there would need to be a node for each carrier variant then but other ones bootloaders might want to select a different revision dtb (who knows how sammy messed with them, can see how many revisions we track in the msm8916 dir) so adding nodes in advance is not something I'd prefer to do...

China forbid carrier lock, but after China have 3 incompatible 3G network for 3 carriers -- UMTS-TDD for CM, UMTS-TDD for CU and CDMA 1xEV-DO for CT, so it's usual to have 3 phone edition. It's not lock.

Reply:
> I think there would need to be a node for each carrier variant then

Just a reminder: if there is indeed such a need, we should probably also distinguish `G900FD` from `G900F*`. They share the same `qcom,msm-id` (`<0xC208FF01 14 0x10000>`) but not the same dts. `F` ([dts](https://github.com/LineageOS/android_kernel_samsung_msm8974/blob/lineage-18.1/arch/arm/boot/dts/msm8974pro/msm8974pro-ac-sec-k-r14.dts)) is a single SIM slot variant and `FD` ([dts](https://github.com/LineageOS/android_kernel_samsung_msm8974/blob/lineage-18.1/arch/arm/boot/dts/msm8974pro/msm8974pro-ac-sec-kltnduos-r14.dts)) is a duo variant.

Reply:
G9000 would've been in line with some 4g devices I've seen, having different networks clarifies it a bit.

so far I didn't see any meaningful difference in single and dual sim device variants, usually either second slot is missing or even present but disabled with nothing that would be important to the OS. (since modem firmware takes care of that) If you know that there is some hardware difference worth exposing to the Linux kernel, we should separate those so lk2nd can differentiate them later.

From my experience, samsung usually has a kernel tree for each device variant and they may have different revisions present so it's not perfect to judge one model by another one's downstream kernel tree. Also sometimes the bootloader picks some other revision than the maximum one... Sometimes even with a wrong SoC id.

Revision here is the second number in the msm-id which matches the one in the downstream dts name. 

IMO ideally lk2nd should be able to distinguish between each and every variant of the device (which doesn't mean that the os it would boot shall as well) so it can hide device-specific quirks. For example one of the plans for lk2nd I have is that is it should be able to pick appropriate Linux device tree on it's own, e.g. it should pick the common klte dts for all, but shall a dedicated dts be needed to one variant, it could be easily added in the future.

Reply:
> G9000 would've been in line with some 4g devices I've seen, having different networks clarifies it a bit.

But... [There do be no such model called `G9000`.](https://www.google.com.hk/search?q=SM-G9000+galaxy+s5) I could only find a clone model from a brand called HDC. Or do you mean `G900\d\w*`?

> Revision here is the second number in the msm-id which matches the one in the downstream dts name.

Yes, I have always kept that in mind. What I mean is that the "revision", the second number in the `msm-id`, actually reflects the hardware revision. I think there couldn't be any argument at that point since what dts describes is exactly the hardware information so different dts means different hardware revision.

> IMO ideally lk2nd should be able to distinguish between each and every variant of the device (which doesn't mean that the os it would boot shall as well) so it can hide device-specific quirks. 

I agree with that, would love and look forward to seeing that. But for now, how many percent of dtb have such an ability? As I have already mentioned, the string matcher needs to be further improved to make that possible, and what's more, filling in an absolutely wrong `match-bootloader` won't break anything. That is, for now, since there is only one dts matches `<0xC208FF01 10 0x10000>`, `match-bootloader` won't take effect. Did I misunderstand it? If this is true, we could probably ignore such unusual bootloader behaviors and get this PR merged first. After lk2nd gains the ability to distinguish strings precisely, time to restrict the `match-bootloader` parameter.

> For example one of the plans for lk2nd I have is that is it should be able to pick appropriate Linux device tree on it's own, e.g. it should pick the common klte dts for all, but shall a dedicated dts be needed to one variant, it could be easily added in the future.

Looking forward to seeing that too. But all `klte` variants has [only one mainline dts](https://github.com/torvalds/linux/blob/master/arch/arm/boot/dts/qcom-msm8974-samsung-klte.dts). What will happen in such a case?

<hr>

<details>

> so far I didn't see any meaningful difference in single and dual sim device variants, usually either second slot is missing or even present but disabled with nothing that would be important to the OS. (since modem firmware takes care of that) If you know that there is some hardware difference worth exposing to the Linux kernel, we should separate those so lk2nd can differentiate them later.
>
> From my experience, samsung usually has a kernel tree for each device variant and they may have different revisions present so it's not perfect to judge one model by another one's downstream kernel tree. Also sometimes the bootloader picks some other revision than the maximum one... Sometimes even with a wrong SoC id.

I see your point. That may be kind of offensive but I must say that these opinions lack self-consistency. Your argument for no need to distinguish single/duo SIM slot(s) variants is "didn't see any meaningful difference", while your argument for the need to distinguish carrier variants is "samsung usually has a kernel tree for each device variant and they may have different revisions present". If the latter is true, that should apply to single/duo SIM slot(s) variants too, so if that is the reason why you want to make lk2nd distinguish carrier variants, no reason that single/duo SIM slot(s) variants may escape.

</details>

Reply:
> There do be no such model called G9000

Right, I just made a guess that there might've been.

>  since there is only one dtb matches <0xC208FF01 10 0x10000>, match-bootloader won't take effect. Did I misunderstand it?

Not really, lk2nd finds the devcie node by the match-... properties. It will still boot but won't know what devcie it's on. 

> But all klte variants has only one mainline dts.

So far :) There were other devices in my "practice" (like Galaxy A5 2015 (a500..)) where one regional variant had slight hardware differences and required a new dts to be introduced.
 
> these opinions lack self-consistency.

Thinking of it, you might be not that wrong. I guess it's right that still I put SS/DS devices as "same" in my mind (as e.g. I've seen the SS device with 2 slots but combo slot being "hidden" with the SD only holder) but the regional ones as "might be different"... It's probably fair to distinguish SS/DS where it makes sense indeed.

fwiw the only thing that I don't particularly like is adding devices that are not tested to work. All other stuff might be relatively easily changed, e.g. when we transition to the different caf tag, we actually consider refactoring samsung related lk2nd dts which might result in squashing everything into a single file per SoC/platform and maybe even squashing the nodes as well. All of that will still need some thought down the line though...

Reply:
I see. Thanks for your detailed replies so far. Now I agree with you.

For now which changes are musts to make this PR get merged? Or do we need to wait for the string matcher to get updated?

Reply:
Thanks for your detailed description. I decided to match `G9006*` and have updated the commits.

In case you need it, [here is the new log](https://github.com/msm8916-mainline/lk2nd/files/8490898/lk_log.txt).